cpp/src/io/json/json_column_csr.cu

```diff
@@ -0,0 +1,360 @@
+    column_categories.begin(),
+    thrust::equal_to<size_type>(),
+    [] __device__(NodeT type_a, NodeT type_b) -> NodeT {
+      auto is_a_leaf = (type_a == NC_VAL || type_a == NC_STR);
+      auto is_b_leaf = (type_b == NC_VAL || type_b == NC_STR);
+      // (v+v=v, *+*=*,  *+v=*, *+#=E, NESTED+VAL=NESTED)
+      // *+*=*, v+v=v
+      if (type_a == type_b) {
+        return type_a;
+      } else if (is_a_leaf) {
+        // *+v=*, N+V=N
+        // STRUCT/LIST + STR/VAL = STRUCT/LIST, STR/VAL + FN = ERR, STR/VAL + STR = STR
+        return type_b == NC_FN ? NC_ERR : (is_b_leaf ? NC_STR : type_b);
+      } else if (is_b_leaf) {
+        return type_a == NC_FN ? NC_ERR : (is_a_leaf ? NC_STR : type_a);
+      }
+      // *+#=E
+      return NC_ERR;
+    });
+
+  auto csr_permutation_it = thrust::make_zip_iterator(
+    thrust::make_permutation_iterator(unique_node_ids.begin(), sorted_column_levels_order.begin()),
+    thrust::make_permutation_iterator(unique_col_ids.begin(), sorted_column_levels_order.begin()),
+    thrust::make_permutation_iterator(max_row_offsets.begin(), sorted_column_levels_order.begin()),
+    thrust::make_permutation_iterator(column_categories.begin(),
+                                      sorted_column_levels_order.begin()));
+  thrust::copy(rmm::exec_policy(stream),
+               csr_permutation_it,
+               csr_permutation_it + num_columns,
+               thrust::make_zip_iterator(csr_unique_node_ids.begin(),
+                                         csr_unique_col_ids.begin(),
+                                         csr_max_row_offsets.begin(),
+                                         csr_column_categories.begin()));
+
+  // 4. unique_copy parent_node_ids, ranges
+  rmm::device_uvector<NodeIndexT> csr_parent_col_ids(num_columns, stream);
+  rmm::device_uvector<SymbolOffsetT> csr_col_range_begin(num_columns, stream);  // Field names
+  rmm::device_uvector<SymbolOffsetT> csr_col_range_end(num_columns, stream);
+  thrust::copy_n(
+    rmm::exec_policy(stream),
+    thrust::make_zip_iterator(
+      thrust::make_permutation_iterator(tree.parent_node_ids.begin(), csr_unique_node_ids.begin()),
+      thrust::make_permutation_iterator(tree.node_range_begin.begin(), csr_unique_node_ids.begin()),
+      thrust::make_permutation_iterator(tree.node_range_end.begin(), csr_unique_node_ids.begin())),
+    csr_unique_node_ids.size(),
+    thrust::make_zip_iterator(
+      csr_parent_col_ids.begin(), csr_col_range_begin.begin(), csr_col_range_end.begin()));
+
+  // convert parent_node_ids to parent_col_ids
+  thrust::transform(
+    rmm::exec_policy(stream),
+    csr_parent_col_ids.begin(),
+    csr_parent_col_ids.end(),
+    csr_parent_col_ids.begin(),
+    [col_ids = original_col_ids.begin()] __device__(auto parent_node_id) -> size_type {
+      return parent_node_id == parent_node_sentinel ? parent_node_sentinel
+                                                    : col_ids[parent_node_id];
+    });
+
```

Comment:
Could we reuse `reduce_to_column_tree` and construct CSR from its result? (trying to reduce code duplication)
do we use any intermediates below?

Reply:
Also, you could consider moving `reduce_to_column_tree` to this file, 
There are 4 stages in json parser.
1. Tokenizer
2. tree construction
3. column construction (type inference, mixed types, pruning and device_column construction)
4. cudf column construction and parsing. (we could consider splitting  parsing as separate too).

Right now, stage 3 & 4 in same json_column.cu. They could be separated.

Reply:
Thanks Karthikeyan, I have moved both the `reduce_to_column_tree` functions to the same file. What do you think about moving implementations for  step 4 in follow-up PRs? #16205 extracts the subtree from the column tree and constructs the cudf column - I'll add the code separation for step 4 there.

Reply:
Also, the reduce calls for `column_categories` and `max_row_offsets` arrays have been merged into one